File: elixir/017/017.exs
defmodule Euler do
  @under_twenty ["zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"]

  @tens ["twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"]

  @denom ["thousand", "million", "billion", "trillion", "quadrillion", "quintillion", "sextillion", "septillion", "octillion", "nonillion", "decillion", "undecillion", "duodecillion", "tredecillion", "quattuordecillion", "sexdecillion", "septendecillion", "octodecillion", "novemdecillion", "vigintillion"]

  def to_english(x) when is_integer(x) and x < 0 do
    "negative " <> to_english(-x)
  end

  def to_english(x) when is_integer(x) and x < 20 do
    Enum.at(@under_twenty, x)
  end

  def to_english(x) when is_integer(x) and x < 100 do
    if rem(x, 10) == 0 do
      Enum.at(@tens, div(x, 10) - 2)
    else
      Enum.at(@tens, div(x, 10) - 2) <> "-" <> Enum.at(@under_twenty, rem(x, 10))
    end
  end

  def to_english(x) when is_integer(x) and x < 1000 do
    result = ""

    d = div(x, 100)

    r = rem(x, 100)

    if d > 0 do
      result = Enum.at(@under_twenty, d) <> " hundred"

      if r > 0 do
        result = result <> " and "
      end
    end

    if r > 0 do
      result = result <> to_english(r)
    end

    result
  end

  def to_english(x) when x < is_integer(x) and x < 1000000000000000000000000000000000000000000000000000000000000000000 do
    denoms(x, 0)
  end

  defp denoms(x, v) when v < length(@denom) do
    didx = v - 1

    dval = round(:math.pow(1000, v))

    if dval > x do
      mod = round(:math.pow(1000, didx))

      l = div(x, mod)

      r = x - (l * mod)

      result = to_english(l) <> " " <> Enum.at(@denom, didx - 1)

      if r > 0 do
        result = result <> ", " <> to_english(r)
      end

      result
    else
      denoms(x, v + 1)
    end
  end
end

IO.puts 1..1000 |> Enum.map(fn x -> Euler.to_english(x) end) |> Enum.join |> :re.replace('\s|-', '', [:global, {:return, :list}]) |> length |> to_string

File: javascript/017/017.js
"use strict";

Number.prototype.toEnglish = function() {
  const underTwenty = ["zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten",
    "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"];

  const tens = ["twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"];

  const denom = ["thousand", "million", "billion", "trillion", "quadrillion", "quintillion", "sextillion", "septillion",
    "octillion", "nonillion", "decillion", "undecillion", "duodecillion", "tredecillion", "quattuordecillion", "sexdecillion",
    "septendecillion", "octodecillion", "novemdecillion", "vigintillion"];

  if (this < 0) {
    return `negative ${(-this).toEnglish()}`;
  } else if (this < 20) {
    return underTwenty[this];
  } else if (this < 100) {
    let result = "";

    for (let v = 0; v < tens.length; v += 1) {
      const cap = tens[v];

      const val = 20 + (10 * v);

      if (val + 10 <= this) {
        continue;
      }

      if (this % 10 != 0) {
        result = `${cap}-${underTwenty[this % 10]}`;
      } else {
        result = cap;
      }

      break;
    }

    return result;
  } else if (this < 1000) {
    let result = "";

    const rem = Math.floor(this / 100);

    const mod = this % 100;

    if (rem > 0) {
      result = `${underTwenty[rem]} hundred`;

      if (mod > 0) {
        result += " and ";
      }
    }

    if (mod > 0) {
      result += mod.toEnglish();
    }

    return result;
  } else if (this < 10 ** 66) {
    let result = "";

    for (let v = 0; v < denom.length; v += 1) {
      const didx = v - 1;

      const dval = 1000 ** v;

      if (dval <= this) {
        continue;
      }

      const mod = 1000 ** didx;

      const l = this / mod;

      const r = this - (l * mod);

      result = `${l.toEnglish()} ${denom[didx - 1]}`;

      if (r > 0) {
        result += `, ${r.toEnglish()}`;
      }

      return result;
    }
  } else {
    return this.toString();
  }
};

let result = "";

for (let i = 1; i <= 1000; i += 1) {
  result += i.toEnglish();
}

console.log(result.replace(/\s|-/g, "").length);
